### PR TITLE
fix(types): far class maker params

### LINKED ISSUE #6718: vat-data initState types assume homogenous array
## Describe the bug

The far class utilities that take `initState` treat the args as a homoegenous array of indeterminate length.

## To Reproduce

Hover to this initState that has exactly one param, a number, and is typed as `...args: (number | undefined)[]`:

Screenshot 2022-12-23 at 12 14 15 PM

## Expected behavior

Parameters are a tuple (which can be heterogeneous). The typedefs should express that. 

## Additional context

One solution may be to make the `init` function be the generic parameter, and use `Parameters<I>` and `ReturnType<I>` instead of the `A` and `S` slots. https://github.com/sindresorhus/type-fest/blob/main/source/set-return-type.d.ts would help.
https://github.com/Agoric/agoric-sdk/blob/9118ab744817d225b9566078bb3b1813e86f0e74/packages/vat-data/src/far-class-utils.js#L25-L28

### PR DESCRIPTION
**Pull request by turadg**
closes: #6718

## Description

see #6718

### Security Considerations

none
### Documentation Considerations

none

### Testing Considerations

New tests

### CODE REVIEW

File: packages/vat-data/src/far-class-utils.js [hunk 1]
```diff
@@ -19,13 +19,14 @@ import {
 
 // TODO interfaceGuard type https://github.com/Agoric/agoric-sdk/issues/6206
 /**
- * @template A,S,T
+ * @template {(...args: any) => any} I init state function
```
Comment (mhofman) [inline packages/vat-data/src/far-class-utils.js:22]: Nit: not strictly necessary, but conventional.
```suggestion
 * @template {(...args: any[]) => any} I init state function
```

**Review by erights — APPROVED** [review]
This is very nice. LGTM!

File: packages/vat-data/src/far-class-utils.js [hunk 2]
```diff
@@ -19,13 +19,14 @@ import {
...
+ * @template T behavior
  * @param {string} tag
  * @param {any} interfaceGuard
- * @param {(...args: A[]) => S} init
- * @param {T & ThisType<{ self: T, state: S }>} methods
- * @param {DefineKindOptions<{ self: T, state: S }>} [options]
- * @returns {(...args: A[]) => (T & RemotableBrand<{}, T>)}
+ * @param {I} init
+ * @param {T & ThisType<{ self: T, state: ReturnType<I> }>} methods
```
Comment (mhofman) [inline packages/vat-data/src/far-class-utils.js:27]: How did I not know about `ThisType` !?

**Review by mhofman — APPROVED** [review]
Looks like I missed the review bandwagon, but LGTM

**Pull request merged**